### PR TITLE
[SPARK-30654][WebUI] Bootstrap4 WebUI upgrade

### PR DESCRIPTION
**Pull request by clarkead**
### What changes were proposed in this pull request?
Spark's Web UI is using an older version of Bootstrap (v. 2.3.2) for the portal pages. Bootstrap 2.x was moved to EOL in Aug 2013 and Bootstrap 3.x was moved to EOL in July 2019 (https://github.com/twbs/release). Older versions of Bootstrap are also getting flagged in security scans for various CVEs:

https://snyk.io/vuln/SNYK-JS-BOOTSTRAP-72889
https://snyk.io/vuln/SNYK-JS-BOOTSTRAP-173700
https://snyk.io/vuln/npm:bootstrap:20180529
https://snyk.io/vuln/npm:bootstrap:20160627

I haven't validated each CVE, but it would be nice to resolve any potential issues and get on a supported release.

The bad news is that there have been quite a few changes between Bootstrap 2 and Bootstrap 4. I've tried updating the library, refactoring/tweaking the CSS and JS to maintain a similar appearance and functionality, and testing the UI for functionality and appearance. This is a fairly large change so I'm sure additional testing and fixes will be needed.

### How was this patch tested?
This has been manually tested, but there is a ton of functionality and there are many pages and detail pages so it is very possible bugs introduced from the upgrade were missed. Additional testing and feedback is welcomed. If it appears a whole page was missed let me know and I'll take a pass at addressing that page/section.

### CODE REVIEW

**Comment by srowen** [comment]
Probably a good idea, yes, but hard to evaluate indeed. Does this update jquery too?
The LICENSE file or licenses/ might need to change just to reflect the new file names.
As far as you know this doesn't change functionality, or doesn't intend to, right?

**Comment by srowen** [comment]
Jenkins test this please

**Comment by clarkead** [comment]
Thank you for taking a look.  The intention with this patch is not to change any functionality.  The goal is to just migrate Spark's WebUI onto a supported Bootstrap release.  Due to Bootstrap changing some of their default CSS the font sizes, paddings, and margins did change.  As a result, I made some overrides in the CSS and some minor changes to the HTML markup to try and maintain the existing feel, but there are still minor differences.  It should be noted that the main navbar does collapse to a toggle button now for mobile devices.  Previously when viewed on a mobile device or small browser window the top navbar would become partially cut off.

jQuery was not modified as it was already on version 3.4.1, but dataTables.bootstrap.min.js was updated to the latest dataTables.bootstrap4.min.js and jquery.dataTables.1.10.18.min.js was updated to jquery.dataTables.min.js (version 1.10.20).  This was mostly needed for compatibility with the latest Bootstrap 4 release.

I noticed this afternoon my initial PR had an issue with my upstream merge so I've added a new commit to resolve it.  I'll look into the reported license issue.

**Comment by srowen** [comment]
Yeah, I do think we need to make a change like this at some point. Thanks for doing the legwork. My only hesitation about merging it soon is that we're about to be in 3.0 code freeze, and it sounds like this may need a little testing. On the one hand, we'll be doing some of that in the 3.0 QA period, so I'm not against it (if you can resolve merge conflicts and feel as confident as you can this works). We can also push back to consider post 3.0.

**Comment by clarkead** [comment]
I've updated the Bootstrap license to reflect the license change to MIT.  I've also updated the dataTables license to reflect the updated date range.  Finally, I had to update the dev/.rat-excludes to reflect the file name changes and additional files provided by the third party libraries.

**Comment by clarkead** [comment]
As far as when to merge it goes, while I'd like to see it merged as part of the 3.0 release I understand the concern regarding the risk it introduces so I'm good if you want to push it to after 3.0 is out the door.  In either case I'll support merge conflicts and will try to resolve visual bug defects found.

As far as my confidence level in the patch, I've tested it for the use cases and pages that I know of, but I'm sure there are areas that I don't know about or missed.  I am confident though that any issues issues found can be quickly resolved.

**Comment by gengliangwang** [comment]
@clarkead Thanks for the work. I am +1 with the upgrade(actually it was in my backlog).
Could you resolve the code conflict? I will help do some manual testing and code review.

File: core/src/main/resources/org/apache/spark/ui/static/executorspage-template.html
Comment (gengliangwang) [inline core/src/main/resources/org/apache/spark/ui/static/executorspage-template.html]: It seems that this file only removes the element ` <div class="container-fluid">` and add `style="width: 100%"`
Could you keep the other part of the file unchanged so that it is easier for review?

**Comment by gengliangwang** [comment]
@clarkead I think we should keep the version in the `dataTable` related file names.

File: core/src/main/resources/org/apache/spark/ui/static/executorspage.js
Comment (gengliangwang) [inline core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: why do we need `ml-4 d-none`

File: licenses/LICENSE-bootstrap.txt
Comment (gengliangwang) [inline licenses/LICENSE-bootstrap.txt]: We need to update `LICENSE-binary` as well

File: core/src/main/resources/org/apache/spark/ui/static/executorspage-template.html
Comment (clarkead) [reply core/src/main/resources/org/apache/spark/ui/static/executorspage-template.html]: Thank you for reviewing the PR.  I just submitted a PR to review the indent formatting updates to make the review easier.

File: core/src/main/resources/org/apache/spark/ui/static/executorspage.js
Comment (clarkead) [reply core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: The ml-4 is needed to add the left margin on the set of checkboxes so that it is indented similarly to the original / current formatting.  Without the ml-4 class the checkboxes are flush with the "Show Additional Metrics" arrow.

With regards to d-none, jQuery 1.8 deprecated .toggle() and jQuery 1.9 removed it so I had to change the JS for the "Show Additional Metrics" expand/collapse functionality.  I decided to change it to the jQuery .toggleClass() method and use the Bootstrap provided d-none (display none) class to show and hide the div instead of creating a new custom class.

Comment (gengliangwang) [reply core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: I see. How about putting the margin-left into the CSS of `container-fluid-div` so that we don't need to write 'ml-4' every time.

File: core/src/main/resources/org/apache/spark/ui/static/executorspage.js
Comment (gengliangwang) [inline core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: Nit: this line seems only adding one space, we can avoid such changes.

File: licenses/LICENSE-bootstrap.txt
Comment (clarkead) [reply licenses/LICENSE-bootstrap.txt]: I just updated the license-binary files associated with bootstrap and datatables.

**Comment by gengliangwang** [comment]
@clarkead I found some issues in the stage page.
This is after changes:
image
The cursor should be a pointer and there should be a list of checkbox after click.
image

File: core/src/main/resources/org/apache/spark/ui/static/executorspage.js
Comment (clarkead) [reply core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: Agreed this is a little lame.  I'll remove the space and use styling instead.

**Comment by clarkead** [comment]
It appears I missed updating the .toggle() calls in the stagepage.js.  It also affects the "Aggregated Metrics by Executor" toggle as well.  I'll resolve that now.

**Comment by gengliangwang** [comment]
There are also some differences in the timeline.
Before changes:
image

After changes:
image

**Comment by clarkead** [comment]
I just pushed a fix right align and the button formatting on the event timeline pagination.

File: core/src/main/resources/org/apache/spark/ui/static/executorspage.js
Comment (clarkead) [reply core/src/main/resources/org/apache/spark/ui/static/executorspage.js]: The space was removed.  It was originally added to make the spacing arrow spacing consistent between the executor page and the other pages since the other pages rely on whitespace for the gap between the arrow and the expand / collapse text.  Fixing all other occurences would require many more code changes so I just reverted the space for now.

File: licenses/LICENSE-bootstrap.txt
Comment (gengliangwang) [reply licenses/LICENSE-bootstrap.txt]: We should also update here:
https://github.com/apache/spark/blob/master/LICENSE-binary#L419

**Comment by gengliangwang** [comment]
@clarkead I noticed that this PR rename the jquery.dataTables.1.10.18.min.js(css) to jquery.dataTables.min.js(css). How about keeping the version in the file name?

**Comment by gengliangwang** [comment]
I manually test this PR change again. I think it is almost ready and will check more closely again before get it merged.
@srowen would you like to take a look at this as well? Also, it seems still a bit risky to put it into 3.0 so I am thinking about making this in master only for now. WDYT?

**Comment by clarkead** [comment]
@gengliangwang I removed the version number as it was at the top of the jquery.dataTables.min.js file and the default file name as provided in the zip did not include the version, but I'm not opposed to adding it back in.  Would you like me to include the version number in the dataTables.bootstrap4.min.js(css) file name as well?

**Comment by gengliangwang** [comment]
@clarkead Yes, let's put the version number in the file name. I don't see the downside of doing so. It makes future maintenance easier.

**Comment by clarkead** [comment]
@gengliangwang I've added the version number to the dataTables js and css filenames as discussed.  I also noticed while testing that the table on the first page of the history server was not displaying at 100% width as it currently does in master so I fixed that via a tweak to core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala (removed an unnecessary "row" class).

**Comment by gengliangwang** [comment]
Jenkins retest this please

**Comment by gengliangwang** [comment]
Jenkins test this please

**Comment by gengliangwang** [comment]
@clarkead this PR is very close to being merged.
Could you fix the test failure?

**Comment by clarkead** [comment]
@gengliangwang I'll take a look at this today and try to get a fix in.

**Comment by gengliangwang** [comment]
add to whitelist

**Comment by gengliangwang** [comment]
Jenkins test this please

**Comment by gengliangwang** [comment]
retest this please.

**Comment by gengliangwang** [comment]
retest this please

**Comment by gengliangwang** [comment]
retest this please

**Review by gengliangwang — APPROVED** [review]
LGTM

**Comment by gengliangwang** [comment]
I will merge this one if no one has other comments tomorrow.

**Comment by gengliangwang** [comment]
Thanks for all the work, merging to master.

**Comment by clarkead** [comment]
@gengliangwang and @srowen Thank you for your time and help getting this merged in.  Much appreciated.

**Comment by emmanuelrflores** [comment]
Has this been put into a release? Looking at [commit](https://github.com/apache/spark/commit/2a4fed0443d6fe066219124833782293630f8a89) by @gengliangwang, it looks like it should be in 3.1.2, but when I look at the release notes for 3.1.2 on the [main site](https://spark.apache.org/releases/spark-release-3-1-2.html) or the [jira site](https://issues.apache.org/jira/secure/ReleaseNote.jspa?projectId=12315420&version=12349602), I can't find the reference. 

I'm not familiar with the release process or how the notes are detailed, so my apologies if I missed something. Thank you

**Comment by srowen** [comment]
The commit message above shows the commit it produces, and that it's in branch 3.1. The JIRA says 3.1.0
https://github.com/apache/spark/commit/2a4fed0443d6fe066219124833782293630f8a89

**Pull request closed without merging**
